snapcraft/internal/lifecycle/_runner.py

```diff
@@ -184,41 +184,56 @@ def run(self, step: steps.Step, part_names=None):
                     # already been built --elopio - 20170713
                     pluginhandler.check_for_collisions(self.config.all_parts)
                 for part in parts:
-                    dirty_report = self._cache.get_dirty_report(
-                        part, current_step)
-                    if dirty_report:
-                        self._handle_dirty(
-                            part, current_step, dirty_report, cli_config)
-                    elif self._cache.has_step_run(part, current_step):
-                        # By default, if a step has already run, don't run it
-                        # again. However, automatically clean and re-run the
-                        # step if all the following conditions apply:
-                        #
-                        #   1. The step is the exact step that was requested
-                        #      (not an earlier one)
-                        #   2. The part was explicitly specified
-                        if (part_names and current_step == step and
-                                part.name in part_names):
-                            getattr(self, '_re{}'.format(
-                                current_step.name))(part)
-                        else:
-                            outdated_report = self._cache.get_outdated_report(
-                                part, current_step)
-                            if outdated_report:
-                                self._handle_outdated(
-                                    part, current_step, outdated_report,
-                                    cli_config)
-                            else:
-                                notify_part_progress(
-                                    part,
-                                    'Skipping {}'.format(current_step.name),
-                                    '(already ran)')
-                    else:
-                        getattr(self, '_run_{}'.format(
-                            current_step.name))(part)
+                    self._handle_step(
+                        part_names, part, step, current_step, cli_config)
 
         self._create_meta(step, processed_part_names)
 
+    def _handle_step(self, requested_part_names: Sequence[str],
+                     part: pluginhandler.PluginHandler,
+                     requested_step: steps.Step,
+                     current_step: steps.Step, cli_config) -> None:
+        # If this step hasn't yet run, all we need to do is run it
+        if not self._cache.has_step_run(part, current_step):
+            getattr(self, '_run_{}'.format(current_step.name))(part)
+            return
```

Comment:
instead of multiple returns here, it seems this could be handled very well with `elif` and a final `else`.

Reply:
We're fetching reports before checking them, which is heavy to do upfront if we don't actually need to, so it doesn't fit particularly well within an if/elif unless we want to nest within `else`s.